# TGIRT smRNA #

This repository stores code for analyzing bias from [mirXplore miRNA reference data](https://www.miltenyibiotec.com/US-en/products/macsmolecular/reagents/nucleic-acid-research/microrna-research/mirxplore-universal-reference.html) sequenced by [TGIRT-seq](http://www.ingex.com/tgirt-enzyme/) using a set of new NTT primer, accompanying the [bioRxiv manuscript](https://www.biorxiv.org/content/early/2018/11/19/474031).

We compared our data to small RNA data generated by [several different small RNA methods](https://www.nature.com/articles/nbt.4183).


## Bias modeling ##

We have seen non-random base distributions within the three nuclotides from each end (5' and 3' ends) from TGIRT [RNA-seq](http://wckdouglas.github.io/assets/article_images/tgirt-seq/seqEnds.png) and [ssDNA-seq (Fig. S5)](https://static-content.springer.com/esm/art%3A10.1038%2Fs41598-017-09064-w/MediaObjects/41598_2017_9064_MOESM1_ESM.pdf). We hypothesized these 6 positions contributes to the sampling bias in small RNA TGIRT-seq. In a TGIRT-seq dataset from miRXplore sample, we demonstrated that the variations from the nucleotides at these positions can explain the over- and under-represented miRNAs (https://github.com/wckdouglas/tgirt_smRNA/blob/master/scripts/RF_correction.ipynb).

Using a random forest regression model, we further modeled how these positions on a miRNA contributed to the sample bias.

$ Y_m = f(x_{m,1},x_{m,2},x_{m,3},x_{m,-3},x_{m,-2},x_{m,-1}) $

where $Y_m$ indicates the difference between observed $log_{10}$CPM and expected $log_{10}$CPM for miRNA $m$. $f$ indicates the random forest regression function, and $x_{m,i}$  indicates the nucleotide of miRNA $m$ at position $i$. Only the first 3 bases ($i = 1$ to 3) and the last 3 bases ($i = -3$ to -1) of each miRNA were considered. Correction of miRNA abundances was done by subtracting $Y_m$ from the experimental $log_{10}$CPM for each miRNA. 
